packages/components/src/box-model-overlay/README.md

```diff
@@ -0,0 +1,158 @@
+# BoxModelOverlay
+
+<div class="callout callout-alert">
+This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
+</div>
+
+`<BoxModelOverlay>` component shows a visual overlay of the [box model](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/The_box_model) (currently only paddings and margins are available) on top of the target element. This is often accompanied by the `<BoxControl>` component to show a preview of the styling changes in the editor.
+
+## Usage
+
+Wrap `<YourComponent>` with `<BoxModelOverlay>` with the `showValues` prop.
+Note that `<YourComponent>` should accept `ref` for `<BoxModelOverlay>` to automatically inject into.
+
+```jsx
+import { __experimentalBoxModelOverlay as BoxModelOverlay } from '@wordpress/components';
+
+// Show all overlays and all sides.
+const showValues = {
+	margin: {
+		top: true,
+		right: true,
+		bottom: true,
+		left: true,
+	},
+	padding: {
+		top: true,
+		right: true,
+		bottom: true,
+		left: true,
+	},
+};
+
+const Example = () => {
+	return (
+		<BoxModelOverlay showValues={ showValues }>
+			<YourComponent />
+		</BoxModelOverlay>
+	);
+};
+```
+
+You can also use the `targetRef` prop to manually pass the ref to `<BoxModelOverlay>` for more advanced usage. This is useful if you need to control where the overlay is rendered or need special handling for the target's `ref`.
+
+```jsx
+const Example = () => {
+	const targetRef = useRef();
+
+	return (
+		<>
+			<YourComponent ref={ targetRef } />
+
+			<BoxModelOverlay
+				showValues={ showValues }
+				targetRef={ targetRef }
+			/>
+		</>
+	);
+};
+```
+
+`<BoxModelOverlay>` internally uses [`Popover`](https://github.com/WordPress/gutenberg/blob/HEAD/packages/components/src/popover/README.md) to position the overlay. This means that you can use `<Popover.Slot>` to alternatively control where the overlay is rendered.
+
+```jsx
+const Example = () => {
+	return (
+		<>
+			<BoxModelOverlay showValues={ showValues }>
+				<YourComponent />
+			</BoxModelOverlay>
+
+			<Popover.Slot />
+		</>
+	);
+};
+```
+
+`<BoxModelOverlay>` under the hood listens to size and style changes of the target element to update the overlay style automatically using `ResizeObserver` and `MutationObserver`. In some edge cases when the observers aren't picking up the changes, you can use the instance method `update` on the ref of the overlay to update it manually.
```

Comment:
Personally, I'd love to see documentation explaining how it works internally so that it feels less _magical_. (Or else the users have to read the source code which might be inaccessible to some given that it's written in TypeScript.)

I don't think changing the documentation would necessary mean breaking changes though. We didn't make any assumptions about how we use them. The purpose of this whole paragraph is even about when they don't work. I'd argue that the pros outweigh the cons if we expose some internal details, but that's just my opinion 😅 .

Reply:
Fair point! Makes sense to me 👍 .